pkg/controller/statusmanager/pod_status.go

```diff
@@ -175,54 +177,67 @@ func (status *StatusManager) updateNodeStatus(nodeName string, ready bool, reaso
 
 
 // Get the pod status from API server
-func (status *StatusManager) SetNodeConditionFromPods(sharedInfo *sharedinfo.SharedInfo) {
+func (status *StatusManager) SetNodeConditionFromPods(podName types.NamespacedName, sharedInfo *sharedinfo.SharedInfo) {
 	status.Lock()
 	defer status.Unlock()
 
-	pods := &corev1.PodList{}
-	err := status.client.List(context.TODO(), pods, client.MatchingLabels{"component": operatortypes.NsxNodeAgentContainerName})
+	now := time.Now()
+	pod := &corev1.Pod{}
+	err := status.client.Get(context.TODO(), podName, pod)
 	if err != nil {
 		log.Error(err, fmt.Sprintf("Error getting %s for node condition", operatortypes.NsxNodeAgentContainerName))
-	} else if len(pods.Items) == 0 {
-		log.Info("nsx-node-agent not found for node condition")
-		nodes := corev1.NodeList{}
-		err = status.client.List(context.TODO(), &nodes)
-		if err != nil {
-			log.Error(err, "Failed to get nodes for condition updating")
-			return
-		}
-		for _, node := range nodes.Items {
-			status.setNodeNetworkUnavailable(node.ObjectMeta.Name, false, "NSXNodeAgent", "Waiting for nsx-node-agent to be created", sharedInfo)
-		}
 		return
 	}
-	pod := corev1.Pod{}
-	for _, pod = range pods.Items {
-		containerStatus := corev1.ContainerStatus{}
-		ready := true
-		messages := ""
-		nodeName := pod.Spec.NodeName
-		for _, containerStatus = range pod.Status.ContainerStatuses {
-			if containerStatus.State.Running == nil {
-				ready = false
-				if containerStatus.State.Waiting != nil {
-					if containerStatus.State.Waiting.Message == "" {
-						messages = messages + fmt.Sprintf("nsx-node-agent/%s not running: %s. ", containerStatus.Name, containerStatus.State.Waiting.Reason)
-					} else {
-						messages = messages + fmt.Sprintf("nsx-node-agent/%s not running: %s. ", containerStatus.Name, containerStatus.State.Waiting.Message)
-					}
-				} else if containerStatus.State.Terminated != nil {
-					if containerStatus.State.Terminated.Message == "" {
-						messages = messages + fmt.Sprintf("nsx-node-agent/%s not running: %s. ", containerStatus.Name, containerStatus.State.Terminated.Reason)
-					} else {
-						messages = messages + fmt.Sprintf("nsx-node-agent/%s not running: %s. ", containerStatus.Name, containerStatus.State.Terminated.Message)
-					}
+	containerStatus := corev1.ContainerStatus{}
+	ready := true
+	// messages is used to store the error message from conatiner status,
+	// tmpMsgs is for the status that pod just restarted in a short time
+	var messages, tmpMsgs string
+	nodeName := pod.Spec.NodeName
+	var startedAt *time.Time
+	var lastStartedAt *time.Time
+	// when pod is pending, its ContainerStatuses is empty
+	if len(pod.Status.ContainerStatuses) == 0 {
```

Comment:
question: can't you just use pod phase for this if condition?

Reply:
According to https://godoc.org/k8s.io/api/core/v1#PodStatus , the pod phase has 5 possible values: Pending, Running, Succeeded, Failed and Unknown. I'm not sure for Failed/Unknown, len(pod.Status.ContainerStatuses) == 0 or > 0, so I think the more insurable method is that if ContainerStatuses is empty, we use the Phase as the error message, no matter the Phase is Pending or other unnormal status.